src/OpenSpaceToolkit/Astrodynamics/Trajectory/Segment.cpp

```diff
@@ -101,6 +105,108 @@ Mass Segment::Solution::computeDeltaMass() const
+    const Shared<Dynamics>& aDynamicsSPtr,
+    const Shared<const Frame>& aFrameSPtr,
+    const Array<Shared<const CoordinatesSubset>>& aCoordinatesSubsetSPtrArray
+) const
+{
+    // Check dynamics is part of the segment dynamics
+    if (!dynamics.contains(aDynamicsSPtr))
+    {
+        throw ostk::core::error::RuntimeError("Provided dynamics is not part of the segment dynamics.");
+    }
+
+    // Extract write coordinates subsets from dynamics
+    const Array<Shared<const CoordinatesSubset>> dynamicsWriteCoordinatesSubsets =
+        aDynamicsSPtr->getWriteCoordinatesSubsets();
+
+    // Check that the provided coordinates subsets are part of the dynamics write coordinates subsets
+    for (auto aCoordinatesSubsetSPtr : aCoordinatesSubsetSPtrArray)
+    {
+        if (!dynamicsWriteCoordinatesSubsets.contains(aCoordinatesSubsetSPtr))
+        {
+            throw ostk::core::error::RuntimeError(
+                "Provided coordinates subset is not part of the dynamics write coordinates subsets."
+            );
+        }
+    }
+
+    // Initialize the definitive coordinate subset array
+    Array<Shared<const CoordinatesSubset>> definitiveCoordinateSubsetArray = aCoordinatesSubsetSPtrArray;
+
+    // Check value for aCoordinatesSubsetSPtrArray
+    if (aCoordinatesSubsetSPtrArray.isEmpty())
+    {
+        definitiveCoordinateSubsetArray = aDynamicsSPtr->getWriteCoordinatesSubsets();
+    }
+
+    // Extract states size
+    const Size stateSize = this->states.getSize();
+
+    // Extract dynamics context and behavior relative to state
+    Array<Shared<const CoordinatesSubset>> dynamicsReadCoordinatesSubsets = aDynamicsSPtr->getReadCoordinatesSubsets();
+
+    // Compute the size of dynamicsContributionMatrix
+    Size dynamicsWriteSize = std::accumulate(
+        definitiveCoordinateSubsetArray.begin(),
+        definitiveCoordinateSubsetArray.end(),
+        0,
+        [](int sum, const Shared<const CoordinatesSubset>& subset)
+        {
+            return sum + subset->getSize();
+        }
+    );
+
+    // Initialize the dynamicsContributionMatrix
+    MatrixXd dynamicsContributionMatrix = MatrixXd::Zero(stateSize, dynamicsWriteSize);
+
+    // Construct the dynamicsContributionMatrix, state by state (a.k.a row by row)
+    for (Index stateIndex = 0; stateIndex < stateSize; ++stateIndex)
+    {
+        const State state = states[stateIndex];
+
+        Shared<const CoordinatesBroker> coordinatesBrokerSPtr = state.accessCoordinatesBroker();
+
+        VectorXd readStateCoordinates =
+            coordinatesBrokerSPtr->extractCoordinates(state.getCoordinates(), dynamicsReadCoordinatesSubsets);
+        VectorXd writeStateCoordinates =
+            coordinatesBrokerSPtr->extractCoordinates(state.getCoordinates(), definitiveCoordinateSubsetArray);
+
+        VectorXd dynamicsContributionAtState =
+            aDynamicsSPtr->computeContribution(state.getInstant(), readStateCoordinates, aFrameSPtr);
```

Comment:
```suggestion
        const VectorXd readStateCoordinates =
            coordinatesBrokerSPtr->extractCoordinates(state.getCoordinates(), dynamicsReadCoordinatesSubsets);
        const VectorXd writeStateCoordinates =
            coordinatesBrokerSPtr->extractCoordinates(state.getCoordinates(), definitiveCoordinateSubsetArray);

        const VectorXd dynamicsContributionAtState =
            aDynamicsSPtr->computeContribution(state.getInstant(), readStateCoordinates, aFrameSPtr);
```

Reply:
Two things:

1. What is `writeStateCoordinates` used for?
2. ⚠️ `computeContribution(state.getInstant(), readStateCoordinates, aFrameSPtr` has not made sure that `readStateCoordinates` are actually resolved in `aFrameSPtr` (at least I could not find it).

Reply:
⚠️ Actually, all-in-all I think it is very much worth it here to just use the `StateBuilder`:

```cpp
const StateBuilder builder = StateBuilder(aFrameSPTr, dynamicsReadCoordinatesSubsets);
```

Then you just use it in the for loop to get the read state coordinates:
```cpp
const VectorXd dynamicsContribution = aDynamicsSPTr->computeContribution(
    state.getInstant(),
    builder.reduce(state.inFrame(aFrameSPTr)), // Consider frame consistency
    aFrameSPTr);
```

I think this is a good use case for the StateBuilder. It will sort thing sort things out for you, warn you if coordinates are not found, if frames are not consistent, etc.

Reply:
1. Agreed, should be removed, not used
2. Great catch, using the StateBuilder suggestion 💯